### ISSUE TITLE
client.FromEnv TLS config is not consistent with the command line tool

### ISSUE THREAD

**Issue by ewxrjk** [moby/moby#39618]
**Description**

The Docker API SDK and the Docker CLI do not treat environment variables related to TLS in the same way.

**Steps to reproduce the issue:**

1. Set up a remote Docker daemon using TLS.
2. Set `DOCKER_HOST=tcp://XXX:2376` and `DOCKER_TLS_VERIFY=1` (per https://docs.docker.com/engine/security/https/). Observe that `docker ps` works.
3. Attempt to connect to the server from a Go program, using `client.NewClientWithOpts(client.FromEnv)`.

**Describe the results you received:**

Actual result: `Error response from daemon: Client sent an HTTP request to an HTTPS server.`

**Describe the results you expected:**

Expected result: successful requests to daemon.

**Additional information you deem important (e.g. issue happens only occasionally):**

For the command line the rules seem to be that TLS is enabled by any of `--tlsverify`, `--tls`, `DOCKER_TLS_VERIFY` or `DOCKER_TLS`.

For `FromEnv`:
* TLS is enabled by setting `DOCKER_CERT_PATH`
* The only effect of `DOCKER_TLS_VERIFY` is to control `InsecureSkipVerify`
* `DOCKER_TLS` is completely ignored.

Since the command line supports the documented configuration, I think it's `FromEnv` that's wrong.

**Output of `docker version`:**

```
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        2d0083d
 Built:             Thu Jun 27 17:57:09 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:20:35 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Containers: 11
 Running: 2
 Paused: 0
 Stopped: 9
Images: 40
Server Version: 19.03.1
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 894b81a4b802e4eb2a91d1ce216b8817763c29fb
runc version: 425e105d5a03fabd737a126ad93d62a9eeede87f
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-9-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 9.769GiB
Name: dekabrach
ID: G5UZ:G2RG:RHQ4:CIZK:OAIE:KUGE:NXFD:J2SK:JIXV:GU6A:JAZ5:ISG2
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by thaJeztah** [comment]
Looks like it's documented in the `fromEnv` option which environment variables are taken into account; https://github.com/moby/moby/blob/28d7dba41d0c0d9c7f0dafcc79d3c59f2b3f5dc3/client/options.go#L19-L57

Other environment variables are handled by the Cli itself, not the client (https://github.com/docker/cli/blob/48bd4c6deb59784b78b860bca14b9995acef0cb3/cli/flags/common.go#L26-L30)

I'm not exactly sure what the history is for that decision, but possibly it's because the flags in which they're used used to be spared between the cli and daemon code (or to show the defaults in the flag description); https://github.com/docker/cli/blob/48bd4c6deb59784b78b860bca14b9995acef0cb3/cli/flags/common.go#L61-L69

Changing the `FromEnv` option to also take those into account would likely be tricky (and may break users that already use this in their Go project), so not sure if we _can_ change this.

@vdemeester @cpuguy83 any ideas?

**Comment by cpuguy83** [comment]
Hmm, I think `FromEnv` should be updated as suggested.
If the envirnonment variables are there it would seem like tls should be enabled.

**Comment by thaJeztah** [comment]
I agree it would be more consistent; do you think it would potentially break users that do not expect this to happen?

**Comment by cpuguy83** [comment]
It's not impossible, but this feels like a bug to me.

On Fri, Aug 9, 2019 at 12:20 PM Sebastiaan van Stijn
<notifications@github.com> wrote:
>
> I agree it would be more consistent; do you think it would potentially break users that do not expect this to happen?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 


- Brian Goff
